packages/react-aria-components/src/utils.tsx

```diff
@@ -87,28 +87,51 @@ interface RenderPropsHookOptions<T> extends RenderProps<T>, SharedDOMProps, Aria
   defaultClassName?: string
 }
 
-export function useRenderProps<T>({className, style, children, defaultClassName, defaultChildren, values, ...otherProps}: RenderPropsHookOptions<T>) {
-  if (typeof className === 'function') {
-    className = className(values);
-  }
+export function useRenderProps<T>(props: RenderPropsHookOptions<T>) {
+  let {
+    className,
+    style,
+    children,
+    defaultClassName,
+    defaultChildren,
+    values,
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    id,
+    ...otherProps
+  } = props;
+
+  return useMemo(() => {
+    let computedClassName: string | undefined;
+    let computedStyle: React.CSSProperties | undefined;
+    let computedChildren: React.ReactNode | undefined;
+
+    if (typeof className === 'function') {
+      computedClassName = className(values);
+    } else {
+      computedClassName = className;
+    }
 
-  if (typeof style === 'function') {
-    style = style(values);
-  }
+    if (typeof style === 'function') {
+      computedStyle = style(values);
+    } else {
+      computedStyle = style;
+    }
 
-  if (typeof children === 'function') {
-    children = children(values);
-  } else if (children == null) {
-    children = defaultChildren;
-  }
+    if (typeof children === 'function') {
+      computedChildren = children(values);
+    } else if (children == null) {
+      computedChildren = defaultChildren;
+    } else {
+      computedChildren = children;
+    }
 
-  delete otherProps.id;
-  return {
-    ...filterDOMProps(otherProps),
-    className: className ?? defaultClassName,
-    style,
-    children
-  };
+    return {
+      ...filterDOMProps(otherProps),
```

Comment:
going to try and replace the filterDOMProps here with an external filterDOMProps call where ever useRenderProps is called so that we doing have to deal with the dep array here.

Also looks like this broke DnD tests for react-aria-components, this may be responsible (DnD was relying on the return object not being memo-ed...)